File: arraySign.java
import java.util.*;
// Approach 1- Counting the negative numbers
class Solution { // TC O(n) SC O(1)
    public int arraySign(int[] nums) {
        int negativeNumber=0;
        for(int num: nums){
            if(num<0){
                negativeNumber++;
            } else if (num==0){
                return 0;
            }
        }
        
        return (negativeNumber%2==0) ? 1:-1;
    }
}

// Approach 2- Flipping the sign
class Solution1{ // TC O(n) SC O(1)
    public int arraySign(int[] nums) {
        int sign = 1;
        for (int num : nums) {
            if (num == 0) {
                return 0;
            }
            if (num < 0) {
                sign = -1 * sign;
            }
        }

        return sign;
    }
}

class arraySign {
    public static void main(String[] args) {
        int[] nums={-1,-2,-3,-4,3,2,1};
        Solution s=new Solution();
        System.out.println(s.arraySign(nums));
        Solution1 s1=new Solution1();
        System.out.println(s1.arraySign(nums));
    }

}

File: arraySign.py
class Solution():
# Approach 1- Counting the negative numbers
    def arraySign(self, nums): # TC O(n) // SC O(1)
        countNegativeNumbers = 0
        for num in nums:
            if num == 0:
                return 0
            if num < 0:
                countNegativeNumbers = countNegativeNumbers + 1

        if countNegativeNumbers %2 == 0:
            return 1
        return -1

# Approach 2- Flipping the sign
    def arraySign1(self, nums): # TC O(n) // SC O(1)
        sign = 1
        for num in nums:
            if num == 0:
                return 0
            if num < 0:
                sign = -1 * sign

        return sign


s=Solution()
print(s.arraySign([-1,-2,-3,-4,3,2,1]))
print(s.arraySign1([-1,-2,-3,-4,3,2,1]))